registry/handlers/context.go

```diff
@@ -15,24 +15,45 @@ import (
 // handlers. Resources that don't need to be shared across handlers should not
 // be on this object.
 type Context struct {
-	// App points to the application structure that created this context.
-	*App
 	context.Context
 
 	// Repository is the repository for the current request. All requests
 	// should be scoped to a single repository. This field may be nil.
 	Repository distribution.Repository
 
-	// Errors is a collection of errors encountered during the request to be
-	// returned to the client API. If errors are added to the collection, the
-	// handler *must not* start the response via http.ResponseWriter.
-	Errors v2.Errors
+	Secret string
 
 	urlBuilder *v2.URLBuilder
+}
+
+// ContextHandler represents an HTTP Handler that also accepts a context object
+// in addition to the ResponseWriter and Request
+type ContextHandler func(*Context, http.ResponseWriter, *http.Request) *HTTPError
+
+// ContextHandlerPartial binds the application to the ContextHandler in a closure
+// such that we can generate request scoped contexts. It also generically handles
+// the various HTTP errors that a ContextHandler might return.
+func ContextHandlerPartial(app *App, ch ContextHandler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		context, err := app.requestContext(w, r)
+		if err == nil {
```

Comment:
Doing it this way let me have the serveJSON code in one place at the bottom of the function. I'll change it if people hate it.